[quant][graphmode][fx] Add qat module mapping support in backend_config_dict

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70152
* __->__ #70287
* #70286

Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33274057](https://our.internmc.facebook.com/intern/diff/D33274057)

jerryzh168 committed:
[quant][graphmode][fx] Add qat module mapping support in backend_config_dict

Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33274057).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add qat module mapping support in backend_config_dict"


Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33274057](https://our.internmc.facebook.com/intern/diff/D33274057)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add qat module mapping support in backend_config_dict"


Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33274057](https://our.internmc.facebook.com/intern/diff/D33274057)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33274057).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add qat module mapping support in backend_config_dict"


Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33274057](https://our.internmc.facebook.com/intern/diff/D33274057)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33274057).

vkuzo commented on torch/ao/quantization/fx/_convert_do_not_use.py:
```diff
@@ -293,6 +293,7 @@ def replace_observer_with_quantize_dequantize_node(graph: Graph, node: Node, mod
                     if isinstance(original_module, torch.nn.intrinsic._FusedModule):
                         fused_module = original_module
                         float_module = fused_module[0]  # type: ignore[index]
+                        print("float module:", float_module)
```
nit: remove?

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -227,10 +228,9 @@ def prepare_get_standalone_module_configs(
-        get_default_qat_module_mappings(), additional_qat_module_mapping)
-    convert(root, mapping=all_mappings, inplace=True, remove_qconfig=False)
+        module_to_qat_module: Dict[Callable, Callable]) -> None:
+    print("module to qat module:", module_to_qat_module)
```
remove?

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1343,7 +1343,12 @@ def prepare(
         additional_qat_module_mapping = prepare_custom_config_dict.get(
             "additional_qat_module_mapping", {})
-        qat_swap_modules(model, additional_qat_module_mapping)
+        if backend_config_dict is None:
```
do we plan to leave this checked in for awhile (weeks/months), or will this be gone soon and backend dict be the only path?

jerryzh168 replied:
this will stay here for a while (a few months) until we fully migrate the convert part of fbgemm/qnnpack to use reference models.

hopefully there is not many people using `additional_qat_module_mapping` api, since we'd like to remove it later. I remember we haven't really talked about this in our tutorials, so it's probably OK

vkuzo replied:
makes sense, maybe we can add a comment to the code with this context?

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add qat module mapping support in backend_config_dict"


Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33274057](https://our.internmc.facebook.com/intern/diff/D33274057)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33274057).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add qat module mapping support in backend_config_dict"


Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33274057](https://our.internmc.facebook.com/intern/diff/D33274057)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33274057).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add qat module mapping support in backend_config_dict"


Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33274057](https://our.internmc.facebook.com/intern/diff/D33274057)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33274057).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add qat module mapping support in backend_config_dict"


Summary:
This PR adds the support to configuring qat modules for fused/non-fused modules
TODO: there are some redundant configs, especially for fused op patterns, we can clean them up later

Test Plan:
python test/fx2trt/test_quant_trt.py TestQuantizeFxTRTOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33274057](https://our.internmc.facebook.com/intern/diff/D33274057)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33274057).
